conmon-rs/server/src/console.rs

```diff
@@ -0,0 +1,119 @@
+            .prefix("conmon-term-")
+            .suffix(".sock")
+            .rand_bytes(7)
+            .tempfile()
+            .context("create tempfile")?;
+        let path: PathBuf = file.path().into();
+        drop(file);
+
+        Ok(Self { path })
+    }
+
+    /// Start the internal listener task and consume ourself
+    pub async fn start(&self) -> Result<()> {
+        let listener = UnixListener::bind(self.path())?;
+        debug!("Listening console socket on {}", self.path().display());
+
+        // Update the permissions
+        let mut perms = fs::metadata(self.path()).await?.permissions();
+        perms.set_mode(0o700);
+        fs::set_permissions(self.path(), perms).await?;
+
+        let (mut stream, _) = listener.accept().await?;
+        debug!("Got console socket stream: {:?}", stream);
+        loop {
+            if stream.ready(Interest::READABLE).await?.is_readable() {
+                let mut data = vec![0; 1024];
+                match stream.try_read(&mut data) {
+                    Ok(n) => {
+                        // Allow only one single read
+                        debug!("Removing socket path {}", self.path().display());
+                        fs::remove_file(self.path()).await?;
+
+                        debug!("Shutting down stream");
+                        stream.shutdown().await?;
+
+                        // TODO: receive file descriptor from socket
+                        info!("Read {} bytes", n);
+
+                        return Ok(());
+                    }
+                    Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
+                        continue;
+                    }
+                    Err(e) => {
+                        return Err(e.into());
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use anyhow::Error;
+    use tokio::net::UnixStream;
+
+    #[tokio::test]
+    async fn start_success() -> Result<()> {
+        let sut = Console::new().await?;
+
+        let path = sut.path().clone();
+        assert!(!path.exists());
+
+        let path_clone = path.clone();
+        let verify = async move {
+            let stream = UnixStream::connect(&path_clone).await?;
+            loop {
+                let ready = stream.ready(Interest::WRITABLE).await?;
+                if ready.is_writable() {
+                    match stream.try_write(b"test") {
+                        Ok(_) => {
+                            return Ok::<(), Error>(());
+                        }
+                        Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
+                            continue;
+                        }
+                        Err(e) => {
+                            return Err(e.into());
+                        }
+                    }
+                }
+            }
+        };
+
+        let (left, right) = tokio::join!(sut.start(), verify);
```

Comment:
@haircommander is this usable? I mean the whole async can run in parallel, but we would have to join the tasks together and the one from the `Console` is blocking on `await`.

Reply:
the goal is to be able to have the container i/o processed in parallel with other i/o processing and container creation. as long as a user's terminal remains responsive, and conmon-rs can make foward progress simultaneously, I think we're good. I am having trouble deciding if this code block does that by looking at it though

Reply:
I'll do some more experiments and unmark the WIP when ready for review.